File: src/main/java/org/cubexmc/railway/util/IntVector3.java
package org.cubexmc.railway.util;

import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.util.Vector;

/**
 * Represents a class containing three immutable integer coordinates: x, y and z
 * Simplified version from BKCommonLib for Railway plugin use
 */
public class IntVector3 implements Comparable<IntVector3> {

    public static final IntVector3 ZERO = new IntVector3(0, 0, 0);
    public final int x, y, z;

    public IntVector3(Block block) {
        this(block.getX(), block.getY(), block.getZ());
    }

    public IntVector3(Location loc) {
        this(loc.getX(), loc.getY(), loc.getZ());
    }

    public IntVector3(final double x, final double y, final double z) {
        this((int) Math.floor(x), (int) Math.floor(y), (int) Math.floor(z));
    }

    public IntVector3(final int x, final int y, final int z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    @Override
    public String toString() {
        return "{" + x + ", " + y + ", " + z + "}";
    }

    @Override
    public boolean equals(Object object) {
        if (object == this) {
            return true;
        } else if (object instanceof IntVector3) {
            IntVector3 other = (IntVector3) object;
            return this.x == other.x && this.y == other.y && this.z == other.z;
        } else {
            return false;
        }
    }

    @Override
    public int hashCode() {
        int hash = 1;
        hash = hash * 31 + this.x;
        hash = hash * 31 + this.y;
        hash = hash * 31 + this.z;
        return hash;
    }

    @Override
    public int compareTo(IntVector3 o) {
        if (this.x != o.x) {
            return Integer.compare(this.x, o.x);
        } else if (this.y != o.y) {
            return Integer.compare(this.y, o.y);
        } else {
            return Integer.compare(this.z, o.z);
        }
    }

    public boolean isSame(IntVector3 other) {
        return this.x == other.x && this.y == other.y && this.z == other.z;
    }

    public IntVector3 add(IntVector3 other) {
        return new IntVector3(this.x + other.x, this.y + other.y, this.z + other.z);
    }

    public IntVector3 subtract(IntVector3 other) {
        return new IntVector3(this.x - other.x, this.y - other.y, this.z - other.z);
    }

    public Block toBlock(World world) {
        return world.getBlockAt(this.x, this.y, this.z);
    }

    public Location toLocation(World world) {
        return new Location(world, this.x, this.y, this.z);
    }

    public Vector toVector() {
        return new Vector(this.x, this.y, this.z);
    }

    public double distance(IntVector3 other) {
        return Math.sqrt(distanceSquared(other));
    }

    public double distanceSquared(IntVector3 other) {
        double dx = this.x - other.x;
        double dy = this.y - other.y;
        double dz = this.z - other.z;
        return dx * dx + dy * dy + dz * dz;
    }
}

File: src/main/java/org/cubexmc/metro/util/IntVector3.kt
package org.cubexmc.metro.util

import org.bukkit.Location
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.util.Vector
import kotlin.math.floor
import kotlin.math.sqrt

/**
 * Represents a class containing three immutable integer coordinates: x, y and z
 * Simplified version from BKCommonLib for Metro plugin use
 */
class IntVector3(@JvmField val x: Int, @JvmField val y: Int, @JvmField val z: Int) : Comparable<IntVector3> {

    constructor(block: Block) : this(block.x, block.y, block.z)

    constructor(loc: Location) : this(loc.x, loc.y, loc.z)

    constructor(x: Double, y: Double, z: Double) : this(
        floor(x).toInt(),
        floor(y).toInt(),
        floor(z).toInt(),
    )

    override fun toString(): String = "{$x, $y, $z}"

    override fun equals(other: Any?): Boolean {
        if (other === this) {
            return true
        }
        return other is IntVector3 && x == other.x && y == other.y && z == other.z
    }

    override fun hashCode(): Int {
        var hash = 1
        hash = hash * HASH_MULTIPLIER + x
        hash = hash * HASH_MULTIPLIER + y
        hash = hash * HASH_MULTIPLIER + z
        return hash
    }

    override fun compareTo(other: IntVector3): Int =
        when {
            x != other.x -> x.compareTo(other.x)
            y != other.y -> y.compareTo(other.y)
            else -> z.compareTo(other.z)
        }

    fun isSame(other: IntVector3): Boolean = x == other.x && y == other.y && z == other.z

    fun add(other: IntVector3): IntVector3 = IntVector3(x + other.x, y + other.y, z + other.z)

    fun subtract(other: IntVector3): IntVector3 = IntVector3(x - other.x, y - other.y, z - other.z)

    fun toBlock(world: World): Block = world.getBlockAt(x, y, z)

    fun toLocation(world: World): Location = Location(world, x.toDouble(), y.toDouble(), z.toDouble())

    fun toVector(): Vector = Vector(x, y, z)

    fun distance(other: IntVector3): Double = sqrt(distanceSquared(other))

    fun distanceSquared(other: IntVector3): Double {
        val dx = (x - other.x).toDouble()
        val dy = (y - other.y).toDouble()
        val dz = (z - other.z).toDouble()
        return dx * dx + dy * dy + dz * dz
    }

    companion object {
        private const val HASH_MULTIPLIER = 31

        @JvmField
        val ZERO = IntVector3(0, 0, 0)
    }
}
